src/string/extglobmatch.c

```diff
@@ -0,0 +1,462 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright © 2014 George Spelvim <linux@horizon.com>
+ * Copyright © 2016 Matthias "ailu" Günzel <a1lu@arcor.de>
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include "extglobmatch.h"
+#include "criterion/internal/common.h"
+
+struct context{
+    int repeat;
+    int except;
+    int nested;
+    char const *pat;
+    char const *str;
+    char const *pat_chunk_end;
+    char const *back_pat;
+    char const *back_str;
+    char const *repeat_pat;
+    char const *repeat_str;
+};
+
+struct ext_glob{
+    char const *cur_pat;
+    char const *end_pat;
+    int ret;
+};
+
+
+static char const *find_pattern_end(char const *pat) {
+    int len = 0;
+    int paren = 0;
+    int patlen = strlen(pat);
+    do {
+        if (*pat == '(')
+            paren++;
+        else if (*pat == ')')
+            paren--;
+        ++len; ++pat;
+    } while (paren > 0 && len < patlen);
+    if (paren) {
+        fprintf(stderr, "pattern error: mismatching parenthesis\n");
+        exit(1);
```

Comment:
1 as an exit status is used to indicate that the tests failed. It could be wiser to indicate invalid running conditions with a value >= 3.